ocaml/xapi/tar_ext.ml

```diff
@@ -0,0 +1,118 @@
+(*
+ * Copyright (C) 2006-2009 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Helpers
+
+module D = Debug.Make (struct let name = "tar_ext" end)
+
+open D
+
+let file_perm = 0o644
+
+let dir_perm = 0o755
+
+let dir_size = 4096L
+
+let ( + ) = Int64.add
+
+exception Illegal_file_path of string
+
+exception Unsupported_file_type of string
+
+exception File_exceeds_max_size_limit of string
+
+exception File_incomplete
+
+exception File_parsed_failed
+
+let unpack dir acc_size max_size ifd head =
+  let assert_file_path_is_legal file_path =
+    if
+      List.exists
+        (fun s -> String.equal "." s || String.equal ".." s)
+        (String.split_on_char '/' file_path)
+    then
+      raise (Illegal_file_path file_path)
+  in
+  let assert_file_not_exceed_max_size acc_size =
+    debug "Current total unpacked file size: %Ld" acc_size ;
+    if Int64.compare acc_size max_size > 0 then
+      raise
+        (File_exceeds_max_size_limit
+           ("File exceeds max size limit: " ^ Int64.to_string max_size)
+        )
+  in
+  let file_name = head.Tar.Header.file_name in
+  debug "Unpacking tar file: %s" file_name ;
+  assert_file_path_is_legal file_name ;
+  let file_path = dir ^ file_name in
+  match head.Tar.Header.link_indicator with
+  | Tar.Header.Link.Normal ->
+      let file_size = head.Tar.Header.file_size in
+      let new_acc_size = acc_size + file_size in
```

Comment:
I see that you use `dir_size` to account for the size of directories. If we do that I think we should also account for storing the filename (perhaps rounded up to 4k).
Otherwise someone could give us a .tar file with billions of 0-sized files, that would still take up space on disk, because we need to store all their filenames (+create an inode).